### ISSUE TITLE
Feedback 55: OIDC Configuration Endpoint URI Structure

### ISSUE THREAD

**Issue by CDR-Register-Stream** [cdr-register/register#55]
The method to derive the path of the OIDC Configuration endpoint is described in the [CDR Register design](https://cdr-register.github.io/register/?#participant-endpoints) however there is ambiguity on what naming conventions should be followed.

The initial premise was that the URI structure as defined by the [Consumer Data Standards](https://consumerdatastandardsaustralia.github.io/standards/#uri-structure) should be followed for consistency.
However, this introduces the following issues:
1. cds-au as the start of the path would infer that the endpoint is for CDR use only. This may not be accurate if the IDP is reused outside of the CDR
2. The \<industry\> keyword may not be appropriate if the Data Holder spans multiple industries and chooses to reuse these IDP endpoints across industries.

~~Another key requirement to be satisfied is to ensure that the major version is contained in the path but NOT in the InfoSecBaseUri, allowing the client to determine the version type to call.~~
*Note in the discussion below, including the version in the path will impact the issuer. Taking this approach would result in tokens that won't be transferable across versions. Therefore, this is no longer a requirement*

Therefore, we propose the following convention options:
1. https://\<InfoSecBaseUri\>/cds-au/v1/infosec/.well-known/openid-configuration
2. https://\<InfoSecBaseUri\>/v1/infosec/.well-known/openid-configuration
3. https://\<InfoSecBaseUri\>/v1/.well-known/openid-configuration
4, https://<InfoSecBaseUri>/.well-known/openid-configuration as is currently defined in the register documentation

The ACCC requests feedback from the community to help validate the above assumptions and identify which naming convention provides enough information but does not impose unnecessary constraints on Data Holders. 

If the current proposed options do not meet this requirement, please propose an alternative.

**Comment by earno** [comment]
OFFICIAL

Nice job Ivan, what’s your preferred option?

**Comment by CDR-Register-Stream** [comment]
@earno In the interest of simplicity and the assumption that an OIDC Configuration endpoint wouldn't be used outside the domain of "infosec", I'd prefer option 3.

However I don't have sufficient insight into the constraints Data Holders may currently have.

**Comment by speedyps** [comment]
As the endpoint is retrieved from the register, this could be a moot point.

However, if we have to have a standard, SISS Data Services preferred option would be to utilise the item 1 alternative

> https://<InfoSecBaseUri>/cds-au/v1/infosec/.well-known/openid-configuration

For clarity, this endpoint is clearly associated with CDR, and provides for the the case where a Data Holder may wish to have other openid-configurations for APIs which are not related to CDR. 

As an aside, what purpose does the V1 and infosec serve?  Could this just be 
https://<InfoSecBaseUri>/cds-au/.well-known/openid-configuration
or 
https://<InfoSecBaseUri>/.well-known/openid-configuration

**Comment by perlboy** [comment]
> The method to derive the path of the OIDC Configuration endpoint is described in the [CDR Register design](https://cdr-register.github.io/register/?#participant-endpoints) however there is ambiguity on what naming conventions should be followed.

Why does it matter? Ultimately InfoSec endpoints are bound by `issuer` during communications which itself is bound to a specific path. If a v2 InfoSec profile alters attributes these would be visible in an altered discovery document.

Specific versions of InfoSec profile can be solved relatively simply by defining `RegisterDataHolderBrandServiceEndpoint`'s URI attributes to be a nested key value set of versions if it is so desired but this would be primarily advisory to recipients although this payload already supports a Standards version.

Beyond human interpretation of the URI, requiring a version lookup to be deterministic based on URI will require either Discovery library modification (breaking existing OIDC Discovery software) or string parsing of the URI (unreliable) and effectively overlaps with the purpose of `openid-configuration`.

While collisions are unlikely, arbitrarily defining a global namespace for a potentially shared service doesn't seem to have much benefit.

> The initial premise was that the URI structure as defined by the [Consumer Data Standards](https://consumerdatastandardsaustralia.github.io/standards/#uri-structure) should be followed for consistency.

A resource server is specific to the resources for which it hosts, an InfoSec server is not which is why it has `client_id`'s whose parameters are negotiated during discovery and registration.

> However, this introduces the following issues:
> 1. cds-au as the start of the path would infer that the endpoint is for CDR use only. This may not be accurate if the IDP is reused outside of the CDR
> 2. The <industry> keyword may not be appropriate if the Data Holder spans multiple industries and chooses to reuse these IDP endpoints across industries.
> Another key requirement to be satisfied is to ensure that the major version is contained in the path but NOT in the InfoSecBaseUri, allowing the client to determine the version type to call.

As stated above. InfoSec discovery occurs via `openid-configuration`. There is no version requirements because the required elements specific to the version are defined in the discovery document. ACCC may wish to advertise that Holder's meet certain versions but this can be easily determined from the discovery document via things such as `claims_supported`, `claims_types_supported` etc.

> Therefore, we propose the following convention options:
> 1. https://<InfoSecBaseUri>/cds-au/v1/infosec/.well-known/openid-configuration
> 2. https://<InfoSecBaseUri>/v1/infosec/.well-known/openid-configuration
> 3. https://<InfoSecBaseUri>/v1/.well-known/openid-configuration

So what is the `issuer`? Is it `InfoSecBaseUri` or the convention? Since `issuer` is bound to tokens and this must match the base_uri from which /.well-known is found, this has flow on effects.

Additionally, how will different brands be handled in this convention? There is a [related discussion about this exact requirement](https://bitbucket.org/openid/fapi/issues/255/certification-clarification-request) and how a single IdP can be used for multiple brands within an organisation. I would suggest ACCC avoids making this more complicated than it already is.

> The ACCC requests feedback from the community to help validate the above assumptions and identify which naming convention provides enough information but does not impose unnecessary constraints on Data Holders.
> If the current proposed options do not meet this requirement, please propose an alternative.

Do nothing. 

_Maybe_ change the Register metadata to include a key value set of `{ 'v1': 'issuer/v1', 'v2':'issuer/v2' }` although `RegisterDataHolderBrandServiceEndpoint` already includes a `version` so this seems superfluous. If a Holder wishes to put different issuer versions at different paths they can do so by advertising a different path (and therefore a different issuer). I flagged this, pre Dynamic Registration decision, [in July 2019](https://github.com/cdr-register/register/issues/4#issuecomment-513405115) stating (in the context of static registration): _While this is likely "future state" scenarios include different auth type per Version or Custom Endpoints_.

If ACCC wishes to determine what Holder's are meeting what versions of the InfoSec profile they can interrogate the discovery document. If this metadata is not available from the discovery document the information should either be registered with IANA so it is available or the specification divergence should be reversed as it will break many installs and no amount of soft-pathing will solve the problem.

In summary, I feel this is solving a "human interpretation" problem for a technical solution which has already been solved much more elegantly with `openid-configuration` and OIDC Discovery Protocol.

**Comment by perlboy** [comment]
>However, if we have to have a standard, SISS Data Services preferred option would be to utilise the item 1 alternative
>https:///cds-au/v1/infosec/.well-known/openid-configuration
> For clarity, this endpoint is clearly associated with CDR, and provides for the the case where a Data Holder may wish to have other openid-configurations for APIs which are not related to CDR.

It also means Data Holder's must now make their `issuer`: https:///cds-au/v1/infosec/

Which means v2 is a different issuer and therefore tokens cannot move between versions.

**Comment by CDR-Register-Stream** [comment]
@perlboy Thanks for highlighting the following:

>  the issuer is bound to tokens and this must match the base_uri from which /.well-known is found

Its good to know that a version numbers within the issuer is not a viable option.

However, what this highlights is Data Holders will need to be sensitive to the value they store in InfosecBaseUri and changes will impact the `issuer`

So introducing option 4, as is currently defined in the [register documentation](https://cdr-register.github.io/register/#participant-endpoints): 
https://\<InfoSecBaseUri\>/.well-known/openid-configuration

**Comment by pcurtisrab** [comment]
I am concerned about a deeper issue underlying this decision regarding MTLS/TLS variability and infrastructure.

Please correct me if I am wrong, but it is not implied by the [Participant Endpoints](https://cdr-register.github.io/register/#participant-endpoints) description that the InfoSecBaseUri covers the OIDC Discovery Configuration Endpoint (TLS), as well as the MTLS auth endpoints (e.g. Token, Revocation)

> Base URI for the Consumer Data Standard InfoSec endpoints. This should encompass all authentication and authorisation related endpoints.

Technically, I suppose a Data Holder may publish their OIDC endpoint there, but in their OIDC configuration point to another host entirely for the various endpoints, not under the InfoSecBaseUri. It seems, however, that this would not be allowed according to the standard. If this assumption is correct, I am concerned that this may pose a problem as Data Holders are required to perform SSL/TLS Renegotiation to support two scenarios on the same host depending on whether or not a client certificate is needed. To my lay understanding, this is problematic from an infrastructure point of view. This is also a side-effect of [removing the MTLS requirement](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/43) on the JWKS and OIDC discovery endpoints.

It seems to me that the InfoSecBaseUri need only point to the OIDC configuration, and need not require endpoints therein to exist under the same URI.

While I have not formed a view regarding a versioning requirement, I otherwise do not see the need for a convention, and would support having the OIDC Dicovery Base URL (defined as the full URL minus the .well-known/openid-configuration) rendered full in the Data Holder metadata.

**Comment by perlboy** [comment]
@pcurtisrab 
> I am concerned about a deeper issue underlying this decision regarding MTLS/TLS variability and infrastructure.
> Please correct me if I am wrong, but it is not implied by the [Participant Endpoints](https://cdr-register.github.io/register/#participant-endpoints) description that the InfoSecBaseUri covers the OIDC Discovery Configuration Endpoint (TLS), as well as the MTLS auth endpoints (e.g. Token, Revocation)

Discovery points to Token/Revocation. As noted by @RalphBragg in the UK there are [multiple hostnames presented](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/95#issuecomment-576640648) specifically because of this split use (Browser facing versus Token claiming). The alternative is to use the Alt-HTTPS port of 8443 with an overlapping certificate which would allow API Gateway's to simply bind to two ports and direct queries where they need to go.

> > Base URI for the Consumer Data Standard InfoSec endpoints. This should encompass all authentication and authorisation related endpoints.
> Technically, I suppose a Data Holder may publish their OIDC endpoint under InfoSecBaseUri, but in their OIDC configuration point to another host entirely for the various endpoints which is not under the InfoSecBaseUri. It seems, however, that this would not be allowed according to the standard. 

Agreed, the statement of "all auth..." is misleading. Perhaps an artefact left over from the original static registration approach? @CDR-Register-Stream perhaps this should be altered to "This should refer to the base uri from which to access the OpenID Discovery document".

>If this assumption is correct, I am concerned that this may pose a problem as Data Holders are required to perform SSL/TLS Renegotiation to support two scenarios on the same host depending on whether or not a client certificate is needed. To my lay understanding, this may be problematic from an infrastructure point of view. This is also a side-effect of [removing the MTLS requirement](https://github.com/ConsumerDataStandardsAustralia/standards-maintenance/issues/43) on the JWKS and OIDC discovery endpoints.

At some point a Holder's Customer's Browser will need to contact the IdP which means that either:
a) The browser will need to see a certificate that is trusted by Global CA's
b) The token claim points are hosted at alternate hostnames
c) The token claim points are hosted at alternate ports with the same hostname 

> It seems to me that the InfoSecBaseUri need only point to the OIDC configuration, and need not require endpoints therein to exist under the same URI.

I agree and I think maybe this is what is actually intended?

**Comment by perlboy** [comment]
> @perlboy Thanks for highlighting the following:
> > the issuer is bound to tokens and this must match the base_uri from which /.well-known is found
> Its good to know that a version numbers within the issuer is not a viable option.

:+1: FWIW it's "viable" but I don't see how it adds any real value because it can only be very rarely changed (ie. You could change it but you could be waiting up to 1 year to wait for existing tokens to expire). Upgrading things in OAuth2/OIDC is solved more by advertising new capabilities then retiring old ones over time, not switching entire endpoint paths.

> However, what this highlights is Data Holders will need to be sensitive to the value they store in InfosecBaseUri and changes will impact the `issuer`

It raises another question about how multiple brands are expected to be presented? Some UK installations currently break spec and there is a reasonable amount of consensus within the WG that the issuer validation is a critical protection for, from memory, CSRF attacks (essentially, if you don't do a === on this you could open yourself up to hijacked token issuance).

> So introducing option 4, as is currently defined in the [register documentation](https://cdr-register.github.io/register/#participant-endpoints):
> https://<InfoSecBaseUri>/.well-known/openid-configuration

:+1:

**Comment by jogu** [comment]
(I'm to some extent repeating points already made, apologies.)

I do not think it is regarded as good practice to put version numbers into the OpenID Connect discovery document path, and it is not something the OpenID/OAuth community has found to be required.

Note that if you do so, the discovery url and the issuer in id_token the server returned to data recipients must be consistent, i.e. using discovery endpoint:

https://<InfoSecBaseUri>/cds-au/v1/infosec/.well-known/openid-configuration

the issuer must be:

https://<InfoSecBaseUri>/cds-au/v1/infosec/

and if the discovery is instead v2:

https://<InfoSecBaseUri>/cds-au/v2/infosec/.well-known/openid-configuration

then the issuer must be:

https://<InfoSecBaseUri>/cds-au/v2/infosec/

(If you don't do this, then you're violating the underlying openid connect discovery standard, and the data recipients may be required to disable security-sensitive checks in the relying party software they choose to use.)

From a practical point of view, having multiple issuers on a single instance of many commercially available openid providers is, at best, tricky to implement - in some/many cases, it is basically infeasible and would cause issues with migration of consents/issued tokens between the different versions, and may cause issues with the stability of pairwise identifiers.


So option 4 seems to make the most sense.

**Comment by anzbankau** [comment]
ANZ's preference is to use https://<InfoSecBaseUri>/cds-au/.well-known/openid-configuration as this aligns with our current interpretation of the standards.

However we are also okay with https://<InfoSecBaseUri>/.well-known/openid-configuration

We do not support the addition of version to the URL.

We also note that our understanding and position is that version in the URL (and x-v header) is not required for the infosec end points.

**Comment by NationalAustraliaBank** [comment]
NAB proposes the InfoSecBaseUri should be equivalent to the OIDC issuer and include the scheme (https://). The current definition of InfoSecBaseUri should be updated as it is somewhat ambiguous. It may be interpreted that all authentication and authorisation endpoints should use this base URI. NAB’s view is the InfoSecBase URI’s purpose is to be used to determine the OIDC configuration endpoint following OIDC discovery rules (https://openid.net/specs/openid-connect-discovery-1_0.html?#ProviderConfig). E.g. <InfoSecBaseUri>/.well-know/openid-configuration

**Comment by WestpacOpenBanking** [comment]
Westpac supports option 4 as described by @CDR-Register-Stream. We support the [arguments](https://github.com/cdr-register/register/issues/55#issuecomment-584985574) made by @perlboy on this matter. Other options may require a change that has potential to impact on our delivery timeframes for July.

**Comment by RalphBragg** [comment]
Option 4 is the most sensible option, the only comment that I will make is that the major commercial providers that are in use by the Australian banks do support multiple issuers by default now with the issuer being restricted to a defined list and selected based on the requested url. This facilitates both multi tenancy of an OP as well as migration changes of the infosecbaseurl. This feature was added to ensure that new implementations of the UK specification were 100 percent discovery spec aligned but had not been rolled out to all participants.

________________________________

**Comment by commbankoss** [comment]
Commonwealth Bank supports option 4. We do not believe there is a benefit to versioning the endpoint. Additionally we agree with @WestpacOpenBanking that further build is required if we do not adopt option 4.

**Comment by CDR-Register-Stream** [comment]
Thank you all for your input. The Register design will not be altered (as per option 4) and a version number will not be incorporated into the path.
However, the documentation will be amended based on the following:

Currently the InfoSecBaseUri endpoint definition is as follows:

> Base URI for the Consumer Data Standard InfoSec endpoints. This should encompass all authentication and authorisation related endpoints

This description is legacy from the old design and doesn't consider the endpoint transport security requirements as described in the [Consumer Data Standards](https://consumerdatastandardsaustralia.github.io/standards/#end-points)

The intent of this endpoint is to provide ADRs a base to discover the OIDC Discovery endpoint as per the [OIDC discovery rules](https://openid.net/specs/openid-connect-discovery-1_0.html?#ProviderConfig). This value must represent the `issuer`

The statement 

> This should encompass all authentication and authorisation related endpoints

will be removed as this constraint is no longer relevant. It will be replaced with reference to the  [OIDC discovery rules](https://openid.net/specs/openid-connect-discovery-1_0.html?#ProviderConfig)

**Comment by CDR-Register-Stream** [comment]
@perlboy @pcurtisrab The above should clarify the intent of the InfoSecBaseUri endpoint. How the Data Holder chooses to break up the different security requirements for different endpoints will not be constrained by the InfoSecBaseUri value.

**Comment by CDR-Register-Stream** [comment]
This issue has now been resolved and documented in release 1.1.1.

Closing as complete
